main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -371,6 +373,195 @@ object Inliner {
+    *   - Returning the rules as is if the rule cannot be determined statically.
+    */
+  @tailrec
+  private def reduceMatch(exp: MonoAst.Expr, rules: List[MonoAst.MatchRule], tpe: Type, eff: Type, loc: SourceLocation)(implicit sym0: Symbol.DefnSym, sctx: SharedContext): Expr = {
+    rules match {
+      case MonoAst.MatchRule(pat, guardOpt, ruleExp) :: rest =>
+        matchRule(exp, pat, guardOpt) match {
+          case MatchResult.Match(binders) =>
+            // Guaranteed rule - convert to let binders.
+            sctx.changed.putIfAbsent(sym0, ())
+            bindPatterns(binders.toSeq, ruleExp, loc)
+          case MatchResult.NoMatch =>
+            // Impossible rule - delete and continue.
+            sctx.changed.putIfAbsent(sym0, ())
+            reduceMatch(exp, rest, tpe, eff, loc)
+          case MatchResult.Unknown =>
+            // Match is unknown - do nothing.
+            Expr.Match(exp, rules, tpe, eff, loc)
+        }
+      case Nil =>
+        // Do nothing.
+        Expr.Match(exp, rules, tpe, eff, loc)
+    }
+  }
+
+  private sealed trait FuzzyBool
+
+  private object FuzzyBool {
+
+    case object True extends FuzzyBool
+
+    case object False extends FuzzyBool
+
+    case object Unknown extends FuzzyBool
+
+  }
+
+  /** Returns the fuzzy evaluation of `exp` as a boolean. */
+  private def evalBoolExpression(exp: MonoAst.Expr): FuzzyBool = exp match {
+    case Expr.Cst(Constant.Bool(true), _, _) => FuzzyBool.True
+    case Expr.Cst(Constant.Bool(false), _, _) => FuzzyBool.False
+    case _ => FuzzyBool.Unknown
+  }
+
+  private sealed trait MatchResult
+
+  private object MatchResult {
+
+    case class Match(binders: Chain[(Option[Pattern.Var], MonoAst.Expr)]) extends MatchResult
+
+    case object NoMatch extends MatchResult
+
+    case object Unknown extends MatchResult
+
+    def emptyMatch(): MatchResult =
+      Match(Chain.empty)
+
+    def singleMatch(pat: Option[Pattern.Var], exp: MonoAst.Expr): MatchResult =
+      Match(Chain((pat, exp)))
+
+    def matchFromBool(b: Boolean): MatchResult =
+      if (b) emptyMatch() else NoMatch
+
+    /**
+      * Concatenates two match results.
+      *
+      * If either result is [[NoMatch]], the output is [[NoMatch]].
+      *
+      * Then if either result is [[Unknown]], the output is [[Unknown]].
+      *
+      * Then (both results now being [[Match]]) the binders are concatenated.
+      */
+    def concat(mr1: MatchResult, mr2: MatchResult): MatchResult = (mr1, mr2) match {
+      case (MatchResult.NoMatch, _) | (_, MatchResult.NoMatch) =>
+        MatchResult.NoMatch
+      case (MatchResult.Unknown, _) | (_, MatchResult.Unknown) =>
+        MatchResult.Unknown
+      case (MatchResult.Match(binders1), MatchResult.Match(binders2)) =>
+        MatchResult.Match(binders1 ++ binders2)
+    }
+
+  }
+
+  /** Returns the match result of `exp` against `pat` with the guard `guardOpt`. */
```

Comment:
This function and the next sounds like the ones where you want an inline code example.